### ISSUE TITLE
Allow extending XHR response error handling globally

### ISSUE THREAD

**Issue by matteocontrini** [flarum/core#2742]
As discussed [here](https://discuss.flarum.org/d/24975-handling-xhr-status-codes-globally), the handling of specific response status codes when doing client XHR requests cannot be currently extended.

It should be possible to show a custom error message depending on the response properties. A possible use-case is to handle a situation where a WAF returned a browser challenge to a API request, and the user wouldn't notice and instead receive a generic error message.

It's probably not a common use-case, so this is low priority and maybe not even a breaking change, but still something that could be improved and that I would find useful.

**Comment by askvortsov1** [comment]
This should be relatively simple, just a matter of rewriting https://github.com/flarum/core/blob/1c4817a0b39a2b1e4a7690e8065df4ef6e1894aa/js/src/common/Application.js#L286-L286 so the error handling logic is in a separate, overridable/extensible method.

**Comment by matteocontrini** [comment]
Still nice to have IMO.

**Comment by matteocontrini** [comment]
So if I wanted to show a specific alert message for a specific status code what should I do now? Override the whole `requestErrorCatch` method? 🤔

**Comment by askvortsov1** [comment]
> So if I wanted to show a specific alert message for a specific status code what should I do now? Override the whole `requestErrorCatch` method? thinking

For now, yes. I'd just check for the error code matching the one you're looking for, and if it's anything else, default to the original implementation.

I suppose at a later time we could refactor this so that current handling is more of a factory and you can add handlers for custom error codes, but I'd like to get a system of JS extenders in place before we start on that.
